// @circlon/angular-tree-component/esm5/lib/components/tree-viewport.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/components/tree-viewport.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, NgZone } from '@angular/core';
import { TreeVirtualScroll } from '../models/tree-virtual-scroll.model';
import { TREE_EVENTS } from '../constants/events';
import { throttle } from 'lodash-es';
var TreeViewportComponent = /** @class */ (function () {
    function TreeViewportComponent(elementRef, ngZone, virtualScroll) {
        var _this = this;
        this.elementRef = elementRef;
        this.ngZone = ngZone;
        this.virtualScroll = virtualScroll;
        this.setViewport = throttle((/**
         * @return {?}
         */
        function () {
            _this.virtualScroll.setViewport(_this.elementRef.nativeElement);
        }), 17);
        this.scrollEventHandler = this.setViewport.bind(this);
    }
    /**
     * @return {?}
     */
    TreeViewportComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.virtualScroll.init();
    };
    /**
     * @return {?}
     */
    TreeViewportComponent.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        setTimeout((/**
         * @return {?}
         */
        function () {
            _this.setViewport();
            _this.virtualScroll.fireEvent({ eventName: TREE_EVENTS.initialized });
        }));
        /** @type {?} */
        var el = this.elementRef.nativeElement;
        this.ngZone.runOutsideAngular((/**
         * @return {?}
         */
        function () {
            el.addEventListener('scroll', _this.scrollEventHandler);
        }));
    };
    /**
     * @return {?}
     */
    TreeViewportComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.virtualScroll.clear();
        /** @type {?} */
        var el = this.elementRef.nativeElement;
        el.removeEventListener('scroll', this.scrollEventHandler);
    };
    /**
     * @return {?}
     */
    TreeViewportComponent.prototype.getTotalHeight = /**
     * @return {?}
     */
    function () {
        return ((this.virtualScroll.isEnabled() &&
            this.virtualScroll.totalHeight + 'px') ||
            'auto');
    };
    TreeViewportComponent.decorators = [
        { type: Component, args: [{
                    selector: 'tree-viewport',
                    providers: [TreeVirtualScroll],
                    template: "\n    <ng-container *treeMobxAutorun=\"{ dontDetach: true }\">\n      <div [style.height]=\"getTotalHeight()\">\n        <ng-content></ng-content>\n      </div>\n    </ng-container>\n  "
                }] }
    ];
    /** @nocollapse */
    TreeViewportComponent.ctorParameters = function () { return [
        { type: ElementRef },
        { type: NgZone },
        { type: TreeVirtualScroll }
    ]; };
    return TreeViewportComponent;
}());
export { TreeViewportComponent };
if (false) {
    /** @type {?} */
    TreeViewportComponent.prototype.setViewport;
    /**
     * @type {?}
     * @private
     */
    TreeViewportComponent.prototype.scrollEventHandler;
    /**
     * @type {?}
     * @private
     */
    TreeViewportComponent.prototype.elementRef;
    /**
     * @type {?}
     * @private
     */
    TreeViewportComponent.prototype.ngZone;
    /** @type {?} */
    TreeViewportComponent.prototype.virtualScroll;
}
