// @circlon/angular-tree-component/esm5/lib/components/tree-node-drop-slot.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/components/tree-node-drop-slot.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, ViewEncapsulation } from '@angular/core';
import { TreeNode } from '../models/tree-node.model';
var TreeNodeDropSlot = /** @class */ (function () {
    function TreeNodeDropSlot() {
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    TreeNodeDropSlot.prototype.onDrop = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        this.node.mouseAction('drop', $event.event, {
            from: $event.element,
            to: { parent: this.node, index: this.dropIndex }
        });
    };
    /**
     * @param {?} element
     * @param {?} $event
     * @return {?}
     */
    TreeNodeDropSlot.prototype.allowDrop = /**
     * @param {?} element
     * @param {?} $event
     * @return {?}
     */
    function (element, $event) {
        return this.node.options.allowDrop(element, { parent: this.node, index: this.dropIndex }, $event);
    };
    TreeNodeDropSlot.decorators = [
        { type: Component, args: [{
                    selector: 'TreeNodeDropSlot, tree-node-drop-slot',
                    encapsulation: ViewEncapsulation.None,
                    template: "\n    <div\n      class=\"node-drop-slot\"\n      (treeDrop)=\"onDrop($event)\"\n      [treeAllowDrop]=\"allowDrop.bind(this)\"\n      [allowDragoverStyling]=\"true\">\n    </div>\n  "
                }] }
    ];
    TreeNodeDropSlot.propDecorators = {
        node: [{ type: Input }],
        dropIndex: [{ type: Input }]
    };
    return TreeNodeDropSlot;
}());
export { TreeNodeDropSlot };
if (false) {
    /** @type {?} */
    TreeNodeDropSlot.prototype.node;
    /** @type {?} */
    TreeNodeDropSlot.prototype.dropIndex;
}
