// @circlon/angular-tree-component/esm5/lib/directives/tree-drop.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/directives/tree-drop.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef, EventEmitter, HostListener, Input, NgZone, Output, Renderer2 } from '@angular/core';
import { TreeDraggedElement } from '../models/tree-dragged-element.model';
/** @type {?} */
var DRAG_OVER_CLASS = 'is-dragging-over';
/** @type {?} */
var DRAG_DISABLED_CLASS = 'is-dragging-over-disabled';
var TreeDropDirective = /** @class */ (function () {
    function TreeDropDirective(el, renderer, treeDraggedElement, ngZone) {
        this.el = el;
        this.renderer = renderer;
        this.treeDraggedElement = treeDraggedElement;
        this.ngZone = ngZone;
        this.allowDragoverStyling = true;
        this.onDropCallback = new EventEmitter();
        this.onDragOverCallback = new EventEmitter();
        this.onDragLeaveCallback = new EventEmitter();
        this.onDragEnterCallback = new EventEmitter();
        this._allowDrop = (/**
         * @param {?} element
         * @param {?} $event
         * @return {?}
         */
        function (element, $event) { return true; });
        this.dragOverEventHandler = this.onDragOver.bind(this);
        this.dragEnterEventHandler = this.onDragEnter.bind(this);
        this.dragLeaveEventHandler = this.onDragLeave.bind(this);
    }
    Object.defineProperty(TreeDropDirective.prototype, "treeAllowDrop", {
        set: /**
         * @param {?} allowDrop
         * @return {?}
         */
        function (allowDrop) {
            if (allowDrop instanceof Function) {
                this._allowDrop = allowDrop;
            }
            else
                this._allowDrop = (/**
                 * @param {?} element
                 * @param {?} $event
                 * @return {?}
                 */
                function (element, $event) { return allowDrop; });
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} $event
     * @return {?}
     */
    TreeDropDirective.prototype.allowDrop = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        return this._allowDrop(this.treeDraggedElement.get(), $event);
    };
    /**
     * @return {?}
     */
    TreeDropDirective.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var el = this.el.nativeElement;
        this.ngZone.runOutsideAngular((/**
         * @return {?}
         */
        function () {
            el.addEventListener('dragover', _this.dragOverEventHandler);
            el.addEventListener('dragenter', _this.dragEnterEventHandler);
            el.addEventListener('dragleave', _this.dragLeaveEventHandler);
        }));
    };
    /**
     * @return {?}
     */
    TreeDropDirective.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var el = this.el.nativeElement;
        el.removeEventListener('dragover', this.dragOverEventHandler);
        el.removeEventListener('dragenter', this.dragEnterEventHandler);
        el.removeEventListener('dragleave', this.dragLeaveEventHandler);
    };
    /**
     * @param {?} $event
     * @return {?}
     */
    TreeDropDirective.prototype.onDragOver = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        if (!this.allowDrop($event)) {
            if (this.allowDragoverStyling) {
                return this.addDisabledClass();
            }
            return;
        }
        this.onDragOverCallback.emit({ event: $event, element: this.treeDraggedElement.get() });
        $event.preventDefault();
        if (this.allowDragoverStyling) {
            this.addClass();
        }
    };
    /**
     * @param {?} $event
     * @return {?}
     */
    TreeDropDirective.prototype.onDragEnter = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        if (!this.allowDrop($event))
            return;
        $event.preventDefault();
        this.onDragEnterCallback.emit({ event: $event, element: this.treeDraggedElement.get() });
    };
    /**
     * @param {?} $event
     * @return {?}
     */
    TreeDropDirective.prototype.onDragLeave = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        if (!this.allowDrop($event)) {
            if (this.allowDragoverStyling) {
                return this.removeDisabledClass();
            }
            return;
        }
        this.onDragLeaveCallback.emit({ event: $event, element: this.treeDraggedElement.get() });
        if (this.allowDragoverStyling) {
            this.removeClass();
        }
    };
    /**
     * @param {?} $event
     * @return {?}
     */
    TreeDropDirective.prototype.onDrop = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        if (!this.allowDrop($event))
            return;
        $event.preventDefault();
        this.onDropCallback.emit({ event: $event, element: this.treeDraggedElement.get() });
        if (this.allowDragoverStyling) {
            this.removeClass();
        }
        this.treeDraggedElement.set(null);
    };
    /**
     * @private
     * @return {?}
     */
    TreeDropDirective.prototype.addClass = /**
     * @private
     * @return {?}
     */
    function () {
        this.renderer.addClass(this.el.nativeElement, DRAG_OVER_CLASS);
    };
    /**
     * @private
     * @return {?}
     */
    TreeDropDirective.prototype.removeClass = /**
     * @private
     * @return {?}
     */
    function () {
        this.renderer.removeClass(this.el.nativeElement, DRAG_OVER_CLASS);
    };
    /**
     * @private
     * @return {?}
     */
    TreeDropDirective.prototype.addDisabledClass = /**
     * @private
     * @return {?}
     */
    function () {
        this.renderer.addClass(this.el.nativeElement, DRAG_DISABLED_CLASS);
    };
    /**
     * @private
     * @return {?}
     */
    TreeDropDirective.prototype.removeDisabledClass = /**
     * @private
     * @return {?}
     */
    function () {
        this.renderer.removeClass(this.el.nativeElement, DRAG_DISABLED_CLASS);
    };
    TreeDropDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[treeDrop]'
                },] }
    ];
    /** @nocollapse */
    TreeDropDirective.ctorParameters = function () { return [
        { type: ElementRef },
        { type: Renderer2 },
        { type: TreeDraggedElement },
        { type: NgZone }
    ]; };
    TreeDropDirective.propDecorators = {
        allowDragoverStyling: [{ type: Input }],
        onDropCallback: [{ type: Output, args: ['treeDrop',] }],
        onDragOverCallback: [{ type: Output, args: ['treeDropDragOver',] }],
        onDragLeaveCallback: [{ type: Output, args: ['treeDropDragLeave',] }],
        onDragEnterCallback: [{ type: Output, args: ['treeDropDragEnter',] }],
        treeAllowDrop: [{ type: Input }],
        onDrop: [{ type: HostListener, args: ['drop', ['$event'],] }]
    };
    return TreeDropDirective;
}());
export { TreeDropDirective };
if (false) {
    /** @type {?} */
    TreeDropDirective.prototype.allowDragoverStyling;
    /** @type {?} */
    TreeDropDirective.prototype.onDropCallback;
    /** @type {?} */
    TreeDropDirective.prototype.onDragOverCallback;
    /** @type {?} */
    TreeDropDirective.prototype.onDragLeaveCallback;
    /** @type {?} */
    TreeDropDirective.prototype.onDragEnterCallback;
    /**
     * @type {?}
     * @private
     */
    TreeDropDirective.prototype.dragOverEventHandler;
    /**
     * @type {?}
     * @private
     */
    TreeDropDirective.prototype.dragEnterEventHandler;
    /**
     * @type {?}
     * @private
     */
    TreeDropDirective.prototype.dragLeaveEventHandler;
    /**
     * @type {?}
     * @private
     */
    TreeDropDirective.prototype._allowDrop;
    /**
     * @type {?}
     * @private
     */
    TreeDropDirective.prototype.el;
    /**
     * @type {?}
     * @private
     */
    TreeDropDirective.prototype.renderer;
    /**
     * @type {?}
     * @private
     */
    TreeDropDirective.prototype.treeDraggedElement;
    /**
     * @type {?}
     * @private
     */
    TreeDropDirective.prototype.ngZone;
}
