// ngx-sortable/__ivy_ngcc__/dist/components/ngx-sortable/ngx-sortable.component.js
import { Component, Input, Output, ContentChild, EventEmitter, TemplateRef } from '@angular/core';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';

function NgxSortableComponent_div_1_Template(rf, ctx) { if (rf & 1) {
    var _r3 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 4);
    ɵngcc0.ɵɵelementStart(1, "label", 5);
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(3, "div", 6);
    ɵngcc0.ɵɵelementStart(4, "button", 7);
    ɵngcc0.ɵɵlistener("click", function NgxSortableComponent_div_1_Template_button_click_4_listener() { ɵngcc0.ɵɵrestoreView(_r3); var ctx_r2 = ɵngcc0.ɵɵnextContext(); return ctx_r2.moveUp(); });
    ɵngcc0.ɵɵtext(5, "\u21E7");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(6, "button", 8);
    ɵngcc0.ɵɵlistener("click", function NgxSortableComponent_div_1_Template_button_click_6_listener() { ɵngcc0.ɵɵrestoreView(_r3); var ctx_r4 = ɵngcc0.ɵɵnextContext(); return ctx_r4.moveDown(); });
    ɵngcc0.ɵɵtext(7, "\u21E9");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(ctx_r0.name);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("disabled", !ctx_r0.selectedItem);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("disabled", !ctx_r0.selectedItem);
} }
function NgxSortableComponent_li_3_div_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "div", 12);
} if (rf & 2) {
    var ctx_r7 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵstyleProp("height", ctx_r7.listStyle.dropZoneHeight);
} }
function NgxSortableComponent_li_3_ng_template_2_Template(rf, ctx) { }
var _c0 = function (a0) { return { "active": a0 }; };
var _c1 = function (a0) { return [a0]; };
function NgxSortableComponent_li_3_Template(rf, ctx) { if (rf & 1) {
    var _r10 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "li", 9);
    ɵngcc0.ɵɵlistener("click", function NgxSortableComponent_li_3_Template_li_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r10); var item_r5 = ctx.$implicit; var ctx_r9 = ɵngcc0.ɵɵnextContext(); return ctx_r9.selectItem(item_r5); })("drop", function NgxSortableComponent_li_3_Template_li_drop_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r10); var i_r6 = ctx.index; var ctx_r11 = ɵngcc0.ɵɵnextContext(); return ctx_r11.onDrop($event, i_r6); })("dragover", function NgxSortableComponent_li_3_Template_li_dragover_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r10); var i_r6 = ctx.index; var ctx_r12 = ɵngcc0.ɵɵnextContext(); return ctx_r12.allowDrop($event, i_r6); })("dragstart", function NgxSortableComponent_li_3_Template_li_dragstart_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r10); var i_r6 = ctx.index; var ctx_r13 = ɵngcc0.ɵɵnextContext(); return ctx_r13.onDragStart($event, i_r6); });
    ɵngcc0.ɵɵtemplate(1, NgxSortableComponent_li_3_div_1_Template, 1, 2, "div", 10);
    ɵngcc0.ɵɵtemplate(2, NgxSortableComponent_li_3_ng_template_2_Template, 0, 0, "ng-template", 11);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var item_r5 = ctx.$implicit;
    var i_r6 = ctx.index;
    var ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(4, _c0, item_r5 == ctx_r1.selectedItem));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r1.onDragOverIndex == i_r6);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", ɵngcc0.ɵɵpureFunction1(6, _c1, item_r5))("ngForTemplate", ctx_r1.itemTemplate);
} }
var NgxSortableComponent = /** @class */ (function () {
    function NgxSortableComponent() {
        this.showHeader = true;
        this.listStyle = {
            height: '250px',
            width: '300px',
            dropZoneHeight: '50px'
        };
        this.listSorted = new EventEmitter();
        this.draggedIndex = -1;
        this.onDragOverIndex = -1;
        console.log('Intializing...');
    }
    NgxSortableComponent.prototype.selectItem = function (item) {
        this.selectedItem = item;
    };
    NgxSortableComponent.prototype.moveUp = function () {
        var index = this.items.indexOf(this.selectedItem);
        if (index === 0) {
            return;
        }
        this.swapElements(index, index - 1);
        this.listSorted.emit(this.items);
    };
    NgxSortableComponent.prototype.moveDown = function () {
        var index = this.items.indexOf(this.selectedItem);
        if (index === this.items.length - 1) {
            return;
        }
        this.swapElements(index, index + 1);
        this.listSorted.emit(this.items);
    };
    NgxSortableComponent.prototype.onDrop = function ($event, index) {
        // index is of the element on which the item is dropped
        this.handleDrop(index);
    };
    NgxSortableComponent.prototype.allowDrop = function ($event, index) {
        // index is of the item on which the item is currently hovered
        this.onDragOverIndex = index;
        $event.preventDefault();
    };
    NgxSortableComponent.prototype.onDragStart = function ($event, index) {
        this.draggedIndex = index;
    };
    NgxSortableComponent.prototype.handleDrop = function (droppedIndex) {
        var item = this.items[this.draggedIndex];
        this.items.splice(this.draggedIndex, 1);
        this.items.splice(droppedIndex, 0, item);
        this.draggedIndex = -1;
        this.onDragOverIndex = -1;
        this.listSorted.emit(this.items);
    };
    NgxSortableComponent.prototype.swapElements = function (oldIndex, newIndex) {
        var temp = this.items[oldIndex];
        this.items[oldIndex] = this.items[newIndex];
        this.items[newIndex] = temp;
    };
    /** @nocollapse */
    NgxSortableComponent.ctorParameters = function () { return []; };
    NgxSortableComponent.propDecorators = {
        'items': [{ type: Input },],
        'name': [{ type: Input },],
        'showHeader': [{ type: Input },],
        'listStyle': [{ type: Input },],
        'listSorted': [{ type: Output },],
        'itemTemplate': [{ type: ContentChild, args: [TemplateRef,] },],
    };
NgxSortableComponent.ɵfac = function NgxSortableComponent_Factory(t) { return new (t || NgxSortableComponent)(); };
NgxSortableComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: NgxSortableComponent, selectors: [["ngx-sortable"]], contentQueries: function NgxSortableComponent_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        ɵngcc0.ɵɵcontentQuery(dirIndex, TemplateRef, true);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.itemTemplate = _t.first);
    } }, inputs: { showHeader: "showHeader", listStyle: "listStyle", items: "items", name: "name" }, outputs: { listSorted: "listSorted" }, decls: 4, vars: 6, consts: [[1, "sortable-container"], ["class", "sortable-header", 4, "ngIf"], [1, "sortable-list"], ["draggable", "true", 3, "ngClass", "click", "drop", "dragover", "dragstart", 4, "ngFor", "ngForOf"], [1, "sortable-header"], [1, "sortable-name"], [1, "sortable-buttons"], ["title", "Move Up", 3, "disabled", "click"], ["title", "Move Down", 3, "disabled", "click"], ["draggable", "true", 3, "ngClass", "click", "drop", "dragover", "dragstart"], ["class", "drop-zone", 3, "height", 4, "ngIf"], ["ngFor", "", 3, "ngForOf", "ngForTemplate"], [1, "drop-zone"]], template: function NgxSortableComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵtemplate(1, NgxSortableComponent_div_1_Template, 8, 3, "div", 1);
        ɵngcc0.ɵɵelementStart(2, "ul", 2);
        ɵngcc0.ɵɵtemplate(3, NgxSortableComponent_li_3_Template, 3, 8, "li", 3);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵstyleProp("width", ctx.listStyle.width);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.showHeader);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵstyleProp("height", ctx.listStyle.height);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngForOf", ctx.items);
    } }, directives: [ɵngcc1.NgIf, ɵngcc1.NgForOf, ɵngcc1.NgClass], styles: [".sortable-container[_ngcontent-%COMP%]{width:100%;border:1px solid #d9d9d9;border-radius:3px}.sortable-container[_ngcontent-%COMP%]   .sortable-header[_ngcontent-%COMP%]{width:auto;padding:5px;background:lightgray;border-bottom:1px solid #d9d9d9;height:30px;background-image:linear-gradient(#f6f7f9 0%, #ebedf0 100%);box-sizing:content-box;color:black}.sortable-container[_ngcontent-%COMP%]   .sortable-header[_ngcontent-%COMP%]   .sortable-name[_ngcontent-%COMP%]{vertical-align:middle;font-size:25px}.sortable-container[_ngcontent-%COMP%]   .sortable-header[_ngcontent-%COMP%]   .sortable-buttons[_ngcontent-%COMP%]{float:right;height:100%;box-sizing:border-box}.sortable-container[_ngcontent-%COMP%]   .sortable-header[_ngcontent-%COMP%]   .sortable-buttons[_ngcontent-%COMP%]   button[_ngcontent-%COMP%]{height:30px;width:30px;vertical-align:middle;display:inline-block;box-sizing:border-box;margin-left:5px;border:1px solid #2399e5;color:#FFFFFF;background:#2399e5;-webkit-transition:background-color .2s;-moz-transition:background-color .2s;transition:background-color .2s;border-radius:5px;outline:none}.sortable-container[_ngcontent-%COMP%]   .sortable-header[_ngcontent-%COMP%]   .sortable-buttons[_ngcontent-%COMP%]   button[_ngcontent-%COMP%]:disabled{cursor:not-allowed}.sortable-container[_ngcontent-%COMP%]   .sortable-header[_ngcontent-%COMP%]   .sortable-buttons[_ngcontent-%COMP%]   button[_ngcontent-%COMP%]:active, .sortable-container[_ngcontent-%COMP%]   .sortable-header[_ngcontent-%COMP%]   .sortable-buttons[_ngcontent-%COMP%]   button[_ngcontent-%COMP%]:hover{border:1px solid #156090;background:#186ba0;color:#ffffff}.sortable-container[_ngcontent-%COMP%]   .sortable-header[_ngcontent-%COMP%]   .sortable-buttons[_ngcontent-%COMP%]   button[_ngcontent-%COMP%]:active:enabled, .sortable-container[_ngcontent-%COMP%]   .sortable-header[_ngcontent-%COMP%]   .sortable-buttons[_ngcontent-%COMP%]   button[_ngcontent-%COMP%]:hover:enabled{cursor:pointer}.sortable-container[_ngcontent-%COMP%]   .sortable-list[_ngcontent-%COMP%]{list-style-type:none;margin:0;padding:0px;overflow:auto}.sortable-container[_ngcontent-%COMP%]   .sortable-list[_ngcontent-%COMP%]   .drop-zone[_ngcontent-%COMP%]{height:30px;background:rgba(225,219,250,0.5);opacity:0.5px;border:2px dotted #e1dbfa}.sortable-container[_ngcontent-%COMP%]   .sortable-list[_ngcontent-%COMP%]   .active[_ngcontent-%COMP%]{background:#3f94e9;color:white}"] });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NgxSortableComponent, [{
        type: Component,
        args: [{
                selector: 'ngx-sortable',
                template: "\n    <div class=\"sortable-container\" [style.width]='listStyle.width'>\n      <div class=\"sortable-header\" *ngIf=\"showHeader\">\n        <label class=\"sortable-name\">{{name}}</label>\n        <div class=\"sortable-buttons\">\n          <button (click)=\"moveUp()\" [disabled]=\"!selectedItem\" title=\"Move Up\">&#8679;</button>\n          <button (click)=\"moveDown()\" [disabled]=\"!selectedItem\" title=\"Move Down\">&#8681;</button>\n        </div>\n      </div>\n      <ul class=\"sortable-list\" [style.height]='listStyle.height'>\n        <li draggable=\"true\" (click)=\"selectItem(item)\" *ngFor=\"let item of items; let i = index;\"\n          [ngClass]=\"{'active': item == selectedItem}\" (drop)=\"onDrop($event, i)\" (dragover)=\"allowDrop($event,i)\"\n          (dragstart)=\"onDragStart($event, i)\">\n          <div class=\"drop-zone\" [style.height]=\"listStyle.dropZoneHeight\" *ngIf=\"onDragOverIndex == i\">\n          </div>\n          <ng-template ngFor [ngForOf]=\"[item]\" [ngForTemplate]=\"itemTemplate\">\n          </ng-template>\n        </li>\n      </ul>\n    </div>\n  ",
                styles: ["\n    .sortable-container{width:100%;border:1px solid #d9d9d9;border-radius:3px}.sortable-container .sortable-header{width:auto;padding:5px;background:lightgray;border-bottom:1px solid #d9d9d9;height:30px;background-image:linear-gradient(#f6f7f9 0%, #ebedf0 100%);box-sizing:content-box;color:black}.sortable-container .sortable-header .sortable-name{vertical-align:middle;font-size:25px}.sortable-container .sortable-header .sortable-buttons{float:right;height:100%;box-sizing:border-box}.sortable-container .sortable-header .sortable-buttons button{height:30px;width:30px;vertical-align:middle;display:inline-block;box-sizing:border-box;margin-left:5px;border:1px solid #2399e5;color:#FFFFFF;background:#2399e5;-webkit-transition:background-color .2s;-moz-transition:background-color .2s;transition:background-color .2s;border-radius:5px;outline:none}.sortable-container .sortable-header .sortable-buttons button:disabled{cursor:not-allowed}.sortable-container .sortable-header .sortable-buttons button:active,.sortable-container .sortable-header .sortable-buttons button:hover{border:1px solid #156090;background:#186ba0;color:#ffffff}.sortable-container .sortable-header .sortable-buttons button:active:enabled,.sortable-container .sortable-header .sortable-buttons button:hover:enabled{cursor:pointer}.sortable-container .sortable-list{list-style-type:none;margin:0;padding:0px;overflow:auto}.sortable-container .sortable-list .drop-zone{height:30px;background:rgba(225,219,250,0.5);opacity:0.5px;border:2px dotted #e1dbfa}.sortable-container .sortable-list .active{background:#3f94e9;color:white}\n  "]
            }]
    }], function () { return []; }, { showHeader: [{
            type: Input
        }], listStyle: [{
            type: Input
        }], listSorted: [{
            type: Output
        }], items: [{
            type: Input
        }], name: [{
            type: Input
        }], itemTemplate: [{
            type: ContentChild,
            args: [TemplateRef]
        }] }); })();
    return NgxSortableComponent;
}());
export { NgxSortableComponent };
