// ngx-easypiechart/fesm5/ngx-easypiechart.js
import { Injectable, Component, Input, ElementRef, NgModule, defineInjectable } from '@angular/core';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
var NgxEasypiechartService = /** @class */ (function () {
    function NgxEasypiechartService() {
    }
    NgxEasypiechartService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] },
    ];
    /** @nocollapse */
    NgxEasypiechartService.ctorParameters = function () { return []; };
    /** @nocollapse */ NgxEasypiechartService.ngInjectableDef = defineInjectable({ factory: function NgxEasypiechartService_Factory() { return new NgxEasypiechartService(); }, token: NgxEasypiechartService, providedIn: "root" });
    return NgxEasypiechartService;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
var NgxEasypiechartComponent = /** @class */ (function () {
    function NgxEasypiechartComponent(element) {
        this.element = element;
    }
    /**
     * @return {?}
     */
    NgxEasypiechartComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @param {?} changes
     * @return {?}
     */
    NgxEasypiechartComponent.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        /** @type {?} */
        var optionsChanged = false;
        if (changes.percent) {
            this.percent = changes.percent.currentValue;
        }
        if (changes.options) {
            this.options = changes.options.currentValue;
            optionsChanged = true;
        }
        this.loadChart(optionsChanged);
    };
    /**
     * @param {?} optionsChanged
     * @return {?}
     */
    NgxEasypiechartComponent.prototype.loadChart = /**
     * @param {?} optionsChanged
     * @return {?}
     */
    function (optionsChanged) {
        if (this.options && this.percent) {
            if (optionsChanged) {
                /** @type {?} */
                var node = this.element.nativeElement;
                // TODO: For some reason , 'EasyPieChart' constructor creates a new chart / canvas everytime it is instantiated.
                // wish there were a method to update options from an existing 'EasyPieChart' instance / object.
                // Until then - we remove all existing children under the given div of the component and then recreate altogether if there are new options
                while (node.hasChildNodes()) {
                    node.removeChild(node.lastChild);
                }
                this.pieChart = new EasyPieChart(node, this.options);
            }
            if (this.pieChart) {
                this.pieChart.update(this.percent);
            }
        }
    };
    NgxEasypiechartComponent.decorators = [
        { type: Component, args: [{
                    selector: 'ngx-easypiechart',
                    template: "<div>Loading ngx-easypiechart...</div>"
                },] },
    ];
    /** @nocollapse */
    NgxEasypiechartComponent.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    NgxEasypiechartComponent.propDecorators = {
        options: [{ type: Input }],
        percent: [{ type: Input }]
    };
    return NgxEasypiechartComponent;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
var NgxEasypiechartModule = /** @class */ (function () {
    function NgxEasypiechartModule() {
    }
    NgxEasypiechartModule.decorators = [
        { type: NgModule, args: [{
                    imports: [],
                    declarations: [NgxEasypiechartComponent],
                    exports: [NgxEasypiechartComponent]
                },] },
    ];
    return NgxEasypiechartModule;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */

export { NgxEasypiechartService, NgxEasypiechartComponent, NgxEasypiechartModule };
