// ngx-sortable/__ivy_ngcc__/dist/ngx-sortable.module.js
import { NgxSortableComponent } from './components/ngx-sortable/ngx-sortable.component';
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import * as ɵngcc0 from '@angular/core';
var NgxSortableModule = /** @class */ (function () {
    function NgxSortableModule() {
    }
    /** @nocollapse */
    NgxSortableModule.ctorParameters = function () { return []; };
NgxSortableModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: NgxSortableModule });
NgxSortableModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function NgxSortableModule_Factory(t) { return new (t || NgxSortableModule)(); }, providers: [], imports: [[CommonModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(NgxSortableModule, { declarations: function () { return [NgxSortableComponent]; }, imports: function () { return [CommonModule]; }, exports: function () { return [NgxSortableComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NgxSortableModule, [{
        type: NgModule,
        args: [{
                declarations: [NgxSortableComponent],
                imports: [CommonModule],
                exports: [NgxSortableComponent],
                providers: [],
                bootstrap: []
            }]
    }], function () { return []; }, null); })();
    return NgxSortableModule;
}());
export { NgxSortableModule };
