// @circlon/angular-tree-component/esm5/lib/mobx-angular/mobx-proxy.js
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spread = (this && this.__spread) || function () {
    for (var ar = [], i = 0; i < arguments.length; i++) ar = ar.concat(__read(arguments[i]));
    return ar;
};
/**
 * @fileoverview added by tsickle
 * Generated from: lib/mobx-angular/mobx-proxy.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { action as mobxAction } from 'mobx';
import { computed as mobxComputed } from 'mobx';
import { observable as mobxObservable } from 'mobx';
// Re-export mobx operators to be able to use inside components with AOT:
/**
 * @param {...?} args
 * @return {?}
 */
export function actionInternal() {
    var args = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        args[_i] = arguments[_i];
    }
    return ((/** @type {?} */ (mobxAction))).apply(void 0, __spread(args));
}
/** @type {?} */
export var action = (/** @type {?} */ (Object.assign(actionInternal, mobxAction)));
/**
 * @param {...?} args
 * @return {?}
 */
function computedInternal() {
    var args = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        args[_i] = arguments[_i];
    }
    return ((/** @type {?} */ (mobxComputed))).apply(void 0, __spread(args));
}
/** @type {?} */
export var computed = (/** @type {?} */ (Object.assign(computedInternal, mobxComputed)));
/**
 * @param {...?} args
 * @return {?}
 */
function observableInternal() {
    var args = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        args[_i] = arguments[_i];
    }
    return ((/** @type {?} */ (mobxObservable))).apply(void 0, __spread(args));
}
/** @type {?} */
export var observable = (/** @type {?} */ (Object.assign(observableInternal, mobxObservable)));
