// @circlon/angular-tree-component/esm5/lib/mobx-angular/tree-mobx-autorun.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/mobx-angular/tree-mobx-autorun.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ViewContainerRef, TemplateRef, Input } from '@angular/core';
import { autorun } from 'mobx';
var TreeMobxAutorunDirective = /** @class */ (function () {
    function TreeMobxAutorunDirective(templateRef, viewContainer) {
        this.templateRef = templateRef;
        this.viewContainer = viewContainer;
        this.templateBindings = {};
    }
    /**
     * @return {?}
     */
    TreeMobxAutorunDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.view = this.viewContainer.createEmbeddedView(this.templateRef);
        if (this.dispose) {
            this.dispose();
        }
        if (this.shouldDetach()) {
            this.view.detach();
        }
        this.autoDetect(this.view);
    };
    /**
     * @return {?}
     */
    TreeMobxAutorunDirective.prototype.shouldDetach = /**
     * @return {?}
     */
    function () {
        return this.treeMobxAutorun && this.treeMobxAutorun.detach;
    };
    /**
     * @param {?} view
     * @return {?}
     */
    TreeMobxAutorunDirective.prototype.autoDetect = /**
     * @param {?} view
     * @return {?}
     */
    function (view) {
        this.dispose = autorun((/**
         * @return {?}
         */
        function () { return view.detectChanges(); }));
    };
    /**
     * @return {?}
     */
    TreeMobxAutorunDirective.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        if (this.dispose) {
            this.dispose();
        }
    };
    TreeMobxAutorunDirective.decorators = [
        { type: Directive, args: [{ selector: '[treeMobxAutorun]' },] }
    ];
    /** @nocollapse */
    TreeMobxAutorunDirective.ctorParameters = function () { return [
        { type: TemplateRef },
        { type: ViewContainerRef }
    ]; };
    TreeMobxAutorunDirective.propDecorators = {
        treeMobxAutorun: [{ type: Input }]
    };
    return TreeMobxAutorunDirective;
}());
export { TreeMobxAutorunDirective };
if (false) {
    /**
     * @type {?}
     * @protected
     */
    TreeMobxAutorunDirective.prototype.templateBindings;
    /**
     * @type {?}
     * @protected
     */
    TreeMobxAutorunDirective.prototype.dispose;
    /**
     * @type {?}
     * @protected
     */
    TreeMobxAutorunDirective.prototype.view;
    /** @type {?} */
    TreeMobxAutorunDirective.prototype.treeMobxAutorun;
    /**
     * @type {?}
     * @protected
     */
    TreeMobxAutorunDirective.prototype.templateRef;
    /**
     * @type {?}
     * @protected
     */
    TreeMobxAutorunDirective.prototype.viewContainer;
}
