// @circlon/angular-tree-component/esm2015/lib/components/tree-node-content.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/components/tree-node-content.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, ViewEncapsulation, TemplateRef } from '@angular/core';
import { TreeNode } from '../models/tree-node.model';
export class TreeNodeContent {
}
TreeNodeContent.decorators = [
    { type: Component, args: [{
                selector: 'tree-node-content',
                encapsulation: ViewEncapsulation.None,
                template: `
  <span *ngIf="!template">{{ node.displayField }}</span>
  <ng-container
    [ngTemplateOutlet]="template"
    [ngTemplateOutletContext]="{ $implicit: node, node: node, index: index }">
  </ng-container>`
            }] }
];
TreeNodeContent.propDecorators = {
    node: [{ type: Input }],
    index: [{ type: Input }],
    template: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    TreeNodeContent.prototype.node;
    /** @type {?} */
    TreeNodeContent.prototype.index;
    /** @type {?} */
    TreeNodeContent.prototype.template;
}
