// ngx-easypiechart/__ivy_ngcc__/fesm2015/ngx-easypiechart.js
import { Injectable, Component, Input, ElementRef, NgModule, defineInjectable } from '@angular/core';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
import * as ɵngcc0 from '@angular/core';
class NgxEasypiechartService {
    constructor() { }
}
NgxEasypiechartService.ɵfac = function NgxEasypiechartService_Factory(t) { return new (t || NgxEasypiechartService)(); };
NgxEasypiechartService.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: NgxEasypiechartService, factory: NgxEasypiechartService.ɵfac, providedIn: 'root' });
/** @nocollapse */
NgxEasypiechartService.ctorParameters = () => [];
/** @nocollapse */ NgxEasypiechartService.ngInjectableDef = defineInjectable({ factory: function NgxEasypiechartService_Factory() { return new NgxEasypiechartService(); }, token: NgxEasypiechartService, providedIn: "root" });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NgxEasypiechartService, [{
        type: Injectable,
        args: [{
                providedIn: 'root'
            }]
    }], function () { return []; }, null); })();

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
class NgxEasypiechartComponent {
    /**
     * @param {?} element
     */
    constructor(element) {
        this.element = element;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        /** @type {?} */
        let optionsChanged = false;
        if (changes.percent) {
            this.percent = changes.percent.currentValue;
        }
        if (changes.options) {
            this.options = changes.options.currentValue;
            optionsChanged = true;
        }
        this.loadChart(optionsChanged);
    }
    /**
     * @param {?} optionsChanged
     * @return {?}
     */
    loadChart(optionsChanged) {
        if (this.options && this.percent) {
            if (optionsChanged) {
                /** @type {?} */
                const node = this.element.nativeElement;
                // TODO: For some reason , 'EasyPieChart' constructor creates a new chart / canvas everytime it is instantiated.
                // wish there were a method to update options from an existing 'EasyPieChart' instance / object.
                // Until then - we remove all existing children under the given div of the component and then recreate altogether if there are new options
                while (node.hasChildNodes()) {
                    node.removeChild(node.lastChild);
                }
                this.pieChart = new EasyPieChart(node, this.options);
            }
            if (this.pieChart) {
                this.pieChart.update(this.percent);
            }
        }
    }
}
NgxEasypiechartComponent.ɵfac = function NgxEasypiechartComponent_Factory(t) { return new (t || NgxEasypiechartComponent)(ɵngcc0.ɵɵdirectiveInject(ElementRef)); };
NgxEasypiechartComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: NgxEasypiechartComponent, selectors: [["ngx-easypiechart"]], inputs: { percent: "percent", options: "options" }, features: [ɵngcc0.ɵɵNgOnChangesFeature], decls: 2, vars: 0, template: function NgxEasypiechartComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div");
        ɵngcc0.ɵɵtext(1, "Loading ngx-easypiechart...");
        ɵngcc0.ɵɵelementEnd();
    } }, encapsulation: 2 });
/** @nocollapse */
NgxEasypiechartComponent.ctorParameters = () => [
    { type: ElementRef }
];
NgxEasypiechartComponent.propDecorators = {
    options: [{ type: Input }],
    percent: [{ type: Input }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NgxEasypiechartComponent, [{
        type: Component,
        args: [{
                selector: 'ngx-easypiechart',
                template: `<div>Loading ngx-easypiechart...</div>`
            }]
    }], function () { return [{ type: ElementRef }]; }, { percent: [{
            type: Input
        }], options: [{
            type: Input
        }] }); })();

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
class NgxEasypiechartModule {
}
NgxEasypiechartModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: NgxEasypiechartModule });
NgxEasypiechartModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function NgxEasypiechartModule_Factory(t) { return new (t || NgxEasypiechartModule)(); }, imports: [[]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(NgxEasypiechartModule, { declarations: [NgxEasypiechartComponent], exports: [NgxEasypiechartComponent] }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NgxEasypiechartModule, [{
        type: NgModule,
        args: [{
                imports: [],
                declarations: [NgxEasypiechartComponent],
                exports: [NgxEasypiechartComponent]
            }]
    }], null, null); })();

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */

export { NgxEasypiechartService, NgxEasypiechartComponent, NgxEasypiechartModule };
