// ng5-breadcrumb/__ivy_ngcc__/src/app/components/breadcrumb/breadcrumb.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { BreadcrumbComponent } from './breadcrumb.component';
import { BreadcrumbService } from './breadcrumb.service';
import * as ɵngcc0 from '@angular/core';
var Ng5BreadcrumbModule = (function () {
    function Ng5BreadcrumbModule() {
    }
    Ng5BreadcrumbModule.forRoot = function () {
        return {
            ngModule: Ng5BreadcrumbModule,
            providers: [BreadcrumbService]
        };
    };
    /** @nocollapse */
    Ng5BreadcrumbModule.ctorParameters = function () { return []; };
Ng5BreadcrumbModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: Ng5BreadcrumbModule });
Ng5BreadcrumbModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function Ng5BreadcrumbModule_Factory(t) { return new (t || Ng5BreadcrumbModule)(); }, imports: [[
            CommonModule
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(Ng5BreadcrumbModule, { declarations: function () { return [BreadcrumbComponent]; }, imports: function () { return [CommonModule]; }, exports: function () { return [BreadcrumbComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(Ng5BreadcrumbModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule
                ],
                exports: [BreadcrumbComponent],
                declarations: [BreadcrumbComponent]
            }]
    }], function () { return []; }, null); })();
    return Ng5BreadcrumbModule;
}());
export { Ng5BreadcrumbModule };
