// @circlon/angular-tree-component/esm2015/lib/mobx-angular/mobx-proxy.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/mobx-angular/mobx-proxy.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { action as mobxAction } from 'mobx';
import { computed as mobxComputed } from 'mobx';
import { observable as mobxObservable } from 'mobx';
// Re-export mobx operators to be able to use inside components with AOT:
/**
 * @param {...?} args
 * @return {?}
 */
export function actionInternal(...args) {
    return ((/** @type {?} */ (mobxAction)))(...args);
}
/** @type {?} */
export const action = (/** @type {?} */ (Object.assign(actionInternal, mobxAction)));
/**
 * @param {...?} args
 * @return {?}
 */
function computedInternal(...args) {
    return ((/** @type {?} */ (mobxComputed)))(...args);
}
/** @type {?} */
export const computed = (/** @type {?} */ (Object.assign(computedInternal, mobxComputed)));
/**
 * @param {...?} args
 * @return {?}
 */
function observableInternal(...args) {
    return ((/** @type {?} */ (mobxObservable)))(...args);
}
/** @type {?} */
export const observable = (/** @type {?} */ (Object.assign(observableInternal, mobxObservable)));
