// ngx-easypiechart/esm2015/lib/ngx-easypiechart.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
import { Component, Input, ElementRef } from '@angular/core';
export class NgxEasypiechartComponent {
    /**
     * @param {?} element
     */
    constructor(element) {
        this.element = element;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        /** @type {?} */
        let optionsChanged = false;
        if (changes.percent) {
            this.percent = changes.percent.currentValue;
        }
        if (changes.options) {
            this.options = changes.options.currentValue;
            optionsChanged = true;
        }
        this.loadChart(optionsChanged);
    }
    /**
     * @param {?} optionsChanged
     * @return {?}
     */
    loadChart(optionsChanged) {
        if (this.options && this.percent) {
            if (optionsChanged) {
                /** @type {?} */
                const node = this.element.nativeElement;
                // TODO: For some reason , 'EasyPieChart' constructor creates a new chart / canvas everytime it is instantiated.
                // wish there were a method to update options from an existing 'EasyPieChart' instance / object.
                // Until then - we remove all existing children under the given div of the component and then recreate altogether if there are new options
                while (node.hasChildNodes()) {
                    node.removeChild(node.lastChild);
                }
                this.pieChart = new EasyPieChart(node, this.options);
            }
            if (this.pieChart) {
                this.pieChart.update(this.percent);
            }
        }
    }
}
NgxEasypiechartComponent.decorators = [
    { type: Component, args: [{
                selector: 'ngx-easypiechart',
                template: `<div>Loading ngx-easypiechart...</div>`
            },] },
];
/** @nocollapse */
NgxEasypiechartComponent.ctorParameters = () => [
    { type: ElementRef }
];
NgxEasypiechartComponent.propDecorators = {
    options: [{ type: Input }],
    percent: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    NgxEasypiechartComponent.prototype.options;
    /** @type {?} */
    NgxEasypiechartComponent.prototype.percent;
    /** @type {?} */
    NgxEasypiechartComponent.prototype.pieChart;
    /** @type {?} */
    NgxEasypiechartComponent.prototype.element;
}
