// ngx-easypiechart/fesm2015/ngx-easypiechart.js
import { Injectable, Component, Input, ElementRef, NgModule, defineInjectable } from '@angular/core';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
class NgxEasypiechartService {
    constructor() { }
}
NgxEasypiechartService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] },
];
/** @nocollapse */
NgxEasypiechartService.ctorParameters = () => [];
/** @nocollapse */ NgxEasypiechartService.ngInjectableDef = defineInjectable({ factory: function NgxEasypiechartService_Factory() { return new NgxEasypiechartService(); }, token: NgxEasypiechartService, providedIn: "root" });

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
class NgxEasypiechartComponent {
    /**
     * @param {?} element
     */
    constructor(element) {
        this.element = element;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        /** @type {?} */
        let optionsChanged = false;
        if (changes.percent) {
            this.percent = changes.percent.currentValue;
        }
        if (changes.options) {
            this.options = changes.options.currentValue;
            optionsChanged = true;
        }
        this.loadChart(optionsChanged);
    }
    /**
     * @param {?} optionsChanged
     * @return {?}
     */
    loadChart(optionsChanged) {
        if (this.options && this.percent) {
            if (optionsChanged) {
                /** @type {?} */
                const node = this.element.nativeElement;
                // TODO: For some reason , 'EasyPieChart' constructor creates a new chart / canvas everytime it is instantiated.
                // wish there were a method to update options from an existing 'EasyPieChart' instance / object.
                // Until then - we remove all existing children under the given div of the component and then recreate altogether if there are new options
                while (node.hasChildNodes()) {
                    node.removeChild(node.lastChild);
                }
                this.pieChart = new EasyPieChart(node, this.options);
            }
            if (this.pieChart) {
                this.pieChart.update(this.percent);
            }
        }
    }
}
NgxEasypiechartComponent.decorators = [
    { type: Component, args: [{
                selector: 'ngx-easypiechart',
                template: `<div>Loading ngx-easypiechart...</div>`
            },] },
];
/** @nocollapse */
NgxEasypiechartComponent.ctorParameters = () => [
    { type: ElementRef }
];
NgxEasypiechartComponent.propDecorators = {
    options: [{ type: Input }],
    percent: [{ type: Input }]
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
class NgxEasypiechartModule {
}
NgxEasypiechartModule.decorators = [
    { type: NgModule, args: [{
                imports: [],
                declarations: [NgxEasypiechartComponent],
                exports: [NgxEasypiechartComponent]
            },] },
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */

export { NgxEasypiechartService, NgxEasypiechartComponent, NgxEasypiechartModule };
